### PR TITLE
Added Dangerfile to root directory of mapknitter

### LINKED ISSUE #303: Add dangerfile to the project
Hi, this is a [first-timers-only issue](https://publiclab.github.io/community-toolbox/#r=all). This means we've worked to make it more legible to folks who either **haven't contributed to our codebase before, or even folks who haven't contributed to open source before**. 

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

We know that the process of creating a pull request is the biggest barrier for new contributors. This issue is for you 💝

If you have contributed before, **consider leaving this one for someone new**, and looking through our general [help wanted](https://github.com/publiclab/mapknitter/labels/help-wanted) issues. Thanks!

### 🤔 What you will need to know.

Nothing. This issue is meant to welcome you to Open Source :) We are happy to walk you through the process.

### 📋 Step by Step

- [ ] 🙋 **Claim this issue**: Comment below. If someone else has claimed it, ask if they've opened a pull request already and if they're stuck -- maybe you can help them solve a problem or move it along!

- [ ] 📝 **Create** the file named [Dangerfile.md] in the `mapknitter` repository (press the little pen Icon) and edit the line as shown below. 

[See this page](https://publiclab.github.io/community-toolbox/#r=all) for some help in taking your first steps!

The file needs to be created at the root of the repository and copy the dangerfile from https://github.com/publiclab/plots2/blob/master/Dangerfile.md

into newly created file. 

- [ ] 💾 **Commit** your changes

- [ ] 🔀 **Start a Pull Request**. There are two ways how you can start a pull request:

1. If you are familiar with the terminal or would like to learn it, [here is a great tutorial](https://egghead.io/series/how-to-contribute-to-an-open-source-project-on-github) on how to send a pull request using the terminal.

2. You can also [edit files directly in your browser](https://help.github.com/articles/editing-files-in-your-repository/) and open a pull request from there. 

- [ ] 🏁 **Done** Ask in comments for a review :)

### 🤔❓ Questions?

Leave a comment below!

### Is someone else already working on this?

We encourage you to link to this issue by mentioning the issue # in your pull request, so we can see if someone's already started on it. **If someone seem stuck, offer them some help!** Otherwise, [take a look at some other issues you can help with](https://publiclab.github.io/community-toolbox/#r=all). Thanks!

### PR DESCRIPTION
**Pull request by satyajeetmaharana**
Fixes #303 

Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [x] tests pass -- `rake test`
* [x] code has been rebased on top of latest master (check if another pull request was added recently, and please rebase)
* [x] pull request is descriptively named and, if possible, multiple commits squashed if they're smaller changes

Thanks!

### CODE REVIEW

File: Dangerfile [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+begin
+
+  junit.parse "output.xml"
+  junit.failures.collect(&:nodes).flatten.each do |failure|
+    failure.nodes.each do |f|
+      f.match(/(test[a-z_\/]+.rb):([0-9]+)/) do |match|
+        source_path = match[1]
+        line = match[2]
+        if !source_path.nil? && !line.nil?
+          f = f.gsub(source_path + ':' + line, "<a href='https://github.com/#{github.pr_author}/plots2/tree/#{github.branch_for_head}/#{source_path}#L#{line}'>#{source_path}:#{line}</a>")
```
Comment (gauravano) [inline Dangerfile]:
```suggestion
          f = f.gsub(source_path + ':' + line, "<a href='https://github.com/#{github.pr_author}/mapknitter/tree/#{github.branch_for_head}/#{source_path}#L#{line}'>#{source_path}:#{line}</a>")
```

File: Dangerfile [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    end
+  end
+
+  junit.errors.collect(&:nodes).flatten.each do |error|
+    error.nodes.each do |f|
+      f.match(/(test[a-z_\/]+.rb):([0-9]+)/) do |match|
+        source_path = match[1]
+        line = match[2]
+        if !source_path.nil? && !line.nil?
+          f = f.gsub(source_path + ':' + line, "<a href='https://github.com/#{github.pr_author}/plots2/tree/#{github.branch_for_head}/#{source_path}#L#{line}'>#{source_path}:#{line}</a>")
```
Comment (gauravano) [inline Dangerfile]:
```suggestion
          f = f.gsub(source_path + ':' + line, "<a href='https://github.com/#{github.pr_author}/mapknitter/tree/#{github.branch_for_head}/#{source_path}#L#{line}'>#{source_path}:#{line}</a>")
```

**Comment by gauravano** [comment]
Please check appropriate boxes in the checklist - https://github.com/publiclab/mapknitter/pull/306#issue-247912338 . Thanks!

**Comment by gauravano** [comment]
There's one more suggestion left. And, just a suggestion: Using Fixes #[issue-number] in the PR comment closes the tagged issue when PR gets merged so it's nice practice. Thanks!

**Comment by satyajeetmaharana** [comment]
Hi @gauravano , Thank for the suggestion. I will keep it as a standard practice in my future PRs. For this issue, I have updated the PR comment. Hope it looks good now. 
Thank you again for your help.

**Comment by gauravano** [comment]
Hey @jywarren, @satyajeetmaharana has added Dangerfile to the repo, could you please complete the installation of dangerbot. Thanks!

**Comment by jywarren** [comment]
Awesome. I think the remaining step is adjusting the `.travis.yml` file and configuring in Travis:

https://danger.systems/guides/getting_started.html#setting-up-danger-to-run-on-your-ci

**Comment by jywarren** [comment]
See for example: https://github.com/publiclab/plots2/blob/master/.travis.yml -- i'll try this.

**Comment by jywarren** [comment]
OK, just noting that `plots2` project has a pretty customized setup. I'll try to get it running in the default mode as in the Danger docs, but if that doesn't work we can copy these lines:

https://github.com/publiclab/plots2/blob/9a2596886e2da54e797905d124d68fac7cddcf58/.travis.yml#L27-L34

**Comment by jywarren** [comment]
I see `command not found: danger` -- so trying the plots2 way!

File: .travis.yml [hunk 3]
```diff
@@ -11,6 +11,10 @@ install:
   - docker-compose run web bash -c "rake db:setup"
   - docker-compose run web bash -c "rake db:migrate"
   - docker-compose run web bower install --allow-root
+  - 'if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then gem install danger danger-junit; fi'
 
 script:
-  - docker-compose run web bash -c "rake test"
+  - docker-compose run web bash -c "CI=TRUE GENERATE_REPORT=true rake test"
```
Comment (jywarren) [inline .travis.yml]: Hmm, oddly `plots2` uses `exec` instead of `run` here. But that didn't work. So trying `run` as it was originally here.

Comment (jywarren) [reply .travis.yml]: `ERROR: No container found for web_1` was the error...

**Comment by jywarren** [comment]
Also adding the correct token to Travis for Danger posting!

**Comment by jywarren** [comment]
Sorry just going to start it again to have it see the token!

**Comment by jywarren** [comment]
Hooray! Looks like it's running, but just not properly parsing the errors produced in JUnit format.

**Comment by jywarren** [comment]
K interesting, it's saying:

```

Don't run Bundler as root. Bundler can ask for sudo if it is needed, and
installing your bundle as root will break this application for all non-root
users on this machine.
Fetching gem metadata from https://rubygems.org/..Retrying dependency api due to error (2/3): Bundler::HTTPError Net::HTTPBadGateway: 
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html>
  <head>
    <title>502 Bad Gateway</title>
  </head>
```

**Comment by jywarren** [comment]
I dunno, i don't see anything obvious so i'm going to restart it again to see if it was some kind of connectivity issue or an upstream server not responding properly?

**Comment by jywarren** [comment]
Looked carefully through plots2 and found we needed additions to test/test_helper.rb!

**Comment by jywarren** [comment]
Trying exec again...

**Comment by alaxalves** [comment]
@jywarren Everytime you use 'run' either with docker or docker-compose it creates a new container, so anything you have done previously won't have desired behavior since you've done it in another container. I suggest changing everywhere that has 'run' to 'exec'.

File: Gemfile [hunk 4]
```diff
@@ -42,6 +42,9 @@ group :test do
 end
 
 group :development do
```
Comment (alaxalves) [inline Gemfile]: This should be in a `group :test` doesn't it? :thinking:

**Comment by jywarren** [comment]
hi @alaxalves - thank you! Would you like to try opening a PR based on this? You know a lot about Docker and we'd LOVE your help!

**Comment by jywarren** [comment]
Or merging this with your PR at #363 ?

**Comment by alaxalves** [comment]
Hey @jywarren I feel like it would be less painful if we get #363 before this. When its done I could open a PR regarding this. What do you think? :smile:

**Comment by jywarren** [comment]
Hahaha, i got the notification from the plotsbot account and responded there, sorry, i thought i was plotsbot! 😂

**Comment by alaxalves** [comment]
@jywarren :laughing: :laughing: :laughing: :laughing: Okay then!!

**Comment by gauravano** [comment]
Hi @alaxalves, as #363 is done. Can we push it to completion? Thank you!

**Comment by kaustubh-nair** [comment]
Reopened in https://github.com/publiclab/mapknitter/pull/926

**Pull request closed without merging**